### PR TITLE
Fix overwrite last written position to zero

### LINKED ISSUE #6049: Incorrect snapshot positions based on invalid last processed position
**Describe the bug**

If we process a record which produces follow up events with positions 1, 2, and 3, the processing state machine's lastWrittenEventPosition will be 3. If we then produce another record which produces NO follow up events, the `writtenEventPosition` will be 0, and the `lastWrittenEventPosition` will also revert to 0. This means we get the following log statements:

```
12:56:48.345 [Broker-0-SnapshotDirector-1] [Broker-0-zb-fs-workers-0] INFO  io.zeebe.logstreams.snapshot - Current commit position 3241 is greater than 0, snapshot 2899-1-1609675008312-3241-9223372036854775807 is valid and has been persisted.
```

This can lead us to mark snapshots as committed too early, since we don't compare the commit position with the correct last written event position.

**To Reproduce**

Start a broker with low segment size (8MB), low snapshot interval (1 minute).

Deploy the standard one task benchmark workflow:

```sh
clients/go/cmd/zbctl/dist/zbctl --insecure deploy benchmarks/project/src/main/resources/bpmn/one_task.bpmn
```

Create workflow instances:

```sh
while true; do clients/go/cmd/zbctl/dist/zbctl --insecure create instance benchmark && sleep 1; done
```

Look at your broker logs and you will see it happen. I'm sure it's easier to write a small test for it though.

**Expected behavior**

If we don't write any follow up events, we should not update the `lastWrittenEventPosition` - it should remain what is the correct last written event position.

**Environment:**
 - OS: Linux
 - Zeebe Version: 0.26.0-SNAPSHOT
 - Configuration: 8MB segment size, 1 minute snapshot interval

### PR DESCRIPTION
**Pull request by miguelpires**
## Description

Fixes an issue where events with no follow-up would overwrite the `lastWrittenEventPosition` to zero which would cause snapshots to be marked as ready before they actually are.

## Related issues

closes #6049 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorTest.java [hunk 1]
```diff
@@ -673,6 +673,55 @@ public void onResumed() {
...
+                    mock(TypedRecordProcessor.class)),
+        onProcessedListener.expect(2));
+
+    // when
+    final long position =
+        streamProcessorRule.writeWorkflowInstanceEvent(WorkflowInstanceIntent.ELEMENT_ACTIVATING);
+    assertThat(onProcessedListener.await()).isTrue();
+    onProcessedListener.expect(1);
+    streamProcessorRule.writeWorkflowInstanceEvent(WorkflowInstanceIntent.CANCEL);
+    assertThat(onProcessedListener.await()).isTrue();
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorTest.java:717]: (optionally) We can separate the two cases to improve the readability. 

```suggestion
    final long position =
        streamProcessorRule.writeWorkflowInstanceEvent(WorkflowInstanceIntent.ELEMENT_ACTIVATING);
    assertThat(onProcessedListener.await()).isTrue();
    
    onProcessedListener.expect(1);
    streamProcessorRule.writeWorkflowInstanceEvent(WorkflowInstanceIntent.CANCEL);
    assertThat(onProcessedListener.await()).isTrue();
```

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorTest.java [hunk 2]
```diff
@@ -673,6 +673,55 @@ public void onResumed() {
...
+
+    // then
+    final long lastWrittenPos =
+        streamProcessorRule.getStreamProcessor(0).getLastWrittenPositionAsync().get();
+    assertThat(lastWrittenPos).isEqualTo(position + 1);
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorTest.java:722]: (optionally) We make the assumption that the position of the follow-up event is increased by `1`. In order to improve the stability of the assertion and to make it more explicit, we could compare the last written position with the position of the `ELEMENT_ACTIVATED` event.

**Review by saig0 — APPROVED** [review]
Looks good :+1: 

I added two minor optional suggestions for the test.

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorTest.java [hunk 3]
```diff
@@ -696,7 +745,7 @@ private AwaitableProcessedListener expect(final int expectedCount) {
     }
 
     private boolean await() throws InterruptedException {
-      return getLatch().await(TIMEOUT.toMillis(), TimeUnit.SECONDS);
+      return getLatch().await(TIMEOUT.toSeconds(), TimeUnit.SECONDS);
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processing/streamprocessor/StreamProcessorTest.java:748]: Good catch :tada:

**Comment by MiguelPires** [comment]
Thanks for the review @saig0, I applied your comments

**Pull request merged**